packages/flutter/lib/src/material/carousel.dart

```diff
@@ -0,0 +1,762 @@
+  // for use when resizing the viewport to non-zero next time.
+  double? _cachedItem;
+
+  @override
+  double? get itemExtent => _itemExtent;
+  double? _itemExtent;
+  set itemExtent(double? value) {
+    if (_itemExtent == value) {
+      return;
+    }
+    _itemExtent = value;
+  }
+
+  double getItemFromPixels(double pixels, double viewportDimension) {
+    assert(viewportDimension > 0.0);
+    final double fraction = itemExtent! / viewportDimension;
+
+    final double actual = math.max(0.0, pixels) / (viewportDimension * fraction);
+    final double round = actual.roundToDouble();
+    if ((actual - round).abs() < precisionErrorTolerance) {
+      return round;
+    }
+    return actual;
+  }
+
+  double getPixelsFromItem(double item) {
+    final double fraction = itemExtent! / viewportDimension;
+
+    return item * viewportDimension * fraction;
+  }
+
+  @override
+  bool applyViewportDimension(double viewportDimension) {
+    final double? oldViewportDimensions = hasViewportDimension ? this.viewportDimension : null;
+    if (viewportDimension == oldViewportDimensions) {
+      return true;
+    }
+    final bool result = super.applyViewportDimension(viewportDimension);
+    final double? oldPixels = hasPixels ? pixels : null;
+    double item;
+    if (oldPixels == null) {
+      item = _itemToShowOnStartup;
+    } else if (oldViewportDimensions == 0.0) {
+      // If resize from zero, we should use the _cachedItem to recover the state.
+      item = _cachedItem!;
+    } else {
+      item = getItemFromPixels(oldPixels, oldViewportDimensions!);
+    }
+    final double newPixels = getPixelsFromItem(item);
+    // If the viewportDimension is zero, cache the page
+    // in case the viewport is resized to be non-zero.
+    _cachedItem = (viewportDimension == 0.0) ? item : null;
+
+    if (newPixels != oldPixels) {
+      correctPixels(newPixels);
+      return false;
+    }
+    return result;
+  }
+
+  @override
+  _CarouselMetrics copyWith({
+    double? minScrollExtent,
+    double? maxScrollExtent,
+    double? pixels,
+    double? viewportDimension,
+    AxisDirection? axisDirection,
+    double? itemExtent,
+    List<int>? layoutWeights,
+    double? devicePixelRatio,
+  }) {
+    return _CarouselMetrics(
+      minScrollExtent: minScrollExtent ?? (hasContentDimensions ? this.minScrollExtent : null),
+      maxScrollExtent: maxScrollExtent ?? (hasContentDimensions ? this.maxScrollExtent : null),
+      pixels: pixels ?? (hasPixels ? this.pixels : null),
+      viewportDimension: viewportDimension ?? (hasViewportDimension ? this.viewportDimension : null),
+      axisDirection: axisDirection ?? this.axisDirection,
+      itemExtent: itemExtent ?? this.itemExtent,
+      devicePixelRatio: devicePixelRatio ?? this.devicePixelRatio,
+    );
+  }
+}
+
+/// A controller for [Carousel].
+///
+/// Using a carousel controller helps to show which item is fully expanded on
+/// the carousel list.
```

Comment:
It looks like it doesn't actually give me this information?

Reply:
Oh this looks a little misleading. Updated. In the second part of Carousel, this controller should be able to show the first fully expanded item.